Added Topic Data Analyzer

irwineffect opened the pull request:
Added a class that subscribes to a topic, listens to data on the topic, then can do various calculations on the received data. Right now it just does min, max, and average, but we can add more advanced statistics later as needed.

Example use case:
For control system testing, we can subscribe to the depth topic and ensure that when we go to depth that we don't have too much overshoot by looking at the minimum depth reached.

irwineffect committed:
started simple class for analyzing data from a topic

people can add more complicated analysis functions as needed.

also added example program showing how to use it.

skallaher commented on src/utility/test_tools.hpp:
```diff
@@ -84,6 +84,118 @@ template <class mtype> class SubscriberTB
+        }
+
+        double min = m_data[0];
+        for (int i = 0; i < m_data.size(); ++i)
```
Since you've already checked the zeroth element, `i` can start at 1. Minor detail.

skallaher commented on src/utility/test_tools.hpp:
```diff
@@ -84,6 +84,118 @@ template <class mtype> class SubscriberTB
+        }
+
+        double max = m_data[0];
+        for (int i = 0; i < m_data.size(); ++i)
```
Ditto here.

skallaher commented on src/utility/test_tools.hpp:
```diff
@@ -84,6 +84,118 @@ template <class mtype> class SubscriberTB
+        }
+
+        double sum = 0.0;
+        for (int i = 0; i < m_data.size(); ++i)
```
Once more.

skallaher reviewed (approved):
Looks good. Just a few little things. Might I suggest a Standard Deviation calculation for future additions.

bkallaher reviewed (changes requested):
Looks good. It would probably be nice to see the live stats. Could you publish the stats to a topic within the analyzer to support this? Having to kill the program to see results seems like a bad example.

irwineffect committed:
changes from review

no need to reiterate over the first element, added calculation of standard deviation
